org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/gtk/OS.d

```diff
@@ -2650,69 +2675,153 @@ public static const int PictOpOver = 3;
 
     static guint GTK_WIDGET_FLAGS( void* arg0 )
     {
+        if (OS.buildVERSION(2, 20, 0) <= OS.GTK_VERSION) {
+            guint flags = 0;
+            if (gtk_widget_is_toplevel(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_TOPLEVEL;
+            if (!gtk_widget_get_has_window(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_NO_WINDOW;
+            if (gtk_widget_get_realized(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_REALIZED;
+            if (gtk_widget_get_mapped(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_MAPPED;
+            if (gtk_widget_get_visible(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_VISIBLE;
+            //if (gtk_widget_is_drawable(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_DRAWABLE;
+            if (gtk_widget_get_sensitive(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_SENSITIVE;
+            if (auto parent = OS.gtk_widget_get_parent(cast(GtkWidget*)arg0)) {
+                if (gtk_widget_get_sensitive(parent)) flags |= GtkWidgetFlags.GTK_PARENT_SENSITIVE;
+            }
+            //if (gtk_widget_is_sensitive(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_IS_SENSITIVE;
+            if (gtk_widget_get_can_focus(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_CAN_FOCUS;
+            if (gtk_widget_has_focus(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_HAS_FOCUS;
+            if (gtk_widget_get_can_default(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_CAN_DEFAULT;
+            if (gtk_widget_get_receives_default(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_RECEIVES_DEFAULT;
+            if (gtk_widget_has_default(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_HAS_DEFAULT;
+            if (gtk_widget_has_grab(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_HAS_GRAB;
+            if (gtk_widget_has_rc_style(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_RC_STYLE;
+            //GValue gvalue;
+            //gtk_widget_style_get_property(cast(GtkWidget*)arg0, "composite-child".ptr, &gvalue);
+            //if (g_value_get_boolean(&gvalue)) flags |= GtkWidgetFlags.GTK_COMPOSITE_CHILD;
+            if (gtk_widget_get_app_paintable(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_APP_PAINTABLE;
+            //if (gtk_widget_get_double_buffered(cast(GtkWidget*)arg0)) flags |= GtkWidgetFlags.GTK_DOUBLE_BUFFERED;
+            return flags;
+        }
         lock.lock();
         scope(exit) lock.unlock();
         return    (cast(GtkObject*) g_type_check_instance_cast (cast(GTypeInstance*) arg0,  gtk_object_get_type ())).flags ;
     }
 
     static ubyte GTK_WIDGET_STATE( void* arg0 )
     {
+        if (OS.buildVERSION(2, 20, 0) <= OS.GTK_VERSION) {
```

Comment:
Are these functions not available on older versions of GTK?

Reply:
Probably works before GTK 2.20, but these is now deprecated.
See also: https://developer.gnome.org/gtk2/stable/GtkWidget.html

Reply:
So do we need to support older versions then?

Reply:
May not be necessary. I don't know if there is a distribution that needs older version.